### PR TITLE
feat: add incremental caching

### PR DESCRIPTION
**Pull request by TimothyMakkison**
Modify `CompilationAnalyzer` to return an equatable model, skipping the final build and emit stage, improving performance and making the IDE more responsive.

- Removed public properties from `CompilationAnalyzer`.
- State is stored in `CompilationModel` converting types into equivalent structurally equatable types.
- Incremental pipeline can cache `CompilationModel` and skip the emit stage in the future if nothing changes.
- Added incremental generator tests.

Note that I had to add `Mediator.SourceGenerator.Roslyn41` to use `WithTrackingName` for the tests. This is pretty fragile as changes to `Mediator.SourceGenerator.Roslyn40` will not be detected by the tests. This could be solved with compile constant preproccessor sections, although I'm not confident I can add this.

### CODE REVIEW

**Comment by martinothamar** [comment]
Awesome work! In my experimental branch I think I removed the Roslyn38 project, was trying to simplify. Do you know what version of the SDK is needed for 4.1?

File: src/Mediator.SourceGenerator.Implementation/Analysis/CompilationAnalyzer.cs [hunk 1]
```diff
@@ -157,7 +108,8 @@ public void Initialize()
         }
         catch (Exception ex)
         {
-            _context.ReportGenericError(ex);
+            Action<Diagnostic> report = ReportDiagnostic;
+            report.ReportGenericError(ex);
```
Comment (martinothamar) [inline src/Mediator.SourceGenerator.Implementation/Analysis/CompilationAnalyzer.cs:112]: I don't quite understand the changes here together with `_diagnostics` and `ReportDiagnostic`. Don't these diagnostic just end up getting dropped on the floor, since they're not sent to the context?

File: src/Mediator.SourceGenerator.Implementation/Analysis/RequestMessageHandlerWrapper.cs [hunk 2]
```diff
@@ -1,33 +1,30 @@
...
+internal sealed record RequestMessageHandlerWrapperModel
 {
-    public readonly string MessageType;
-    private readonly CompilationAnalyzer _analyzer;
-
-    public RequestMessageHandlerWrapper(string messageType, CompilationAnalyzer analyzer)
+    public RequestMessageHandlerWrapperModel(string messageType, CompilationAnalyzer analyzer)
     {
         MessageType = messageType;
-        _analyzer = analyzer;
```
Comment (martinothamar) [inline src/Mediator.SourceGenerator.Implementation/Analysis/RequestMessageHandlerWrapper.cs:13]: I think this change breaks cases where the namespace is configured by the user. `TryParseConfiguration` is the last step of analysis for some reason, at which point `MediatorNamespace` as already been captured here. We could move up `TryParseConfiguration`, maybe that's cleaner. What do you think?

File: src/Mediator.SourceGenerator.Implementation/Models/RequestMessageHandlerWrapperModel.cs [hunk 3]
```diff
@@ -0,0 +1,7 @@
+﻿namespace Mediator.SourceGenerator;
+
+internal sealed record RequestMessageHandlerWrapperModel2
+{
+    // TODO delete
+    public void WeirdBug() => throw new Exception();
```
Comment (martinothamar) [inline src/Mediator.SourceGenerator.Implementation/Models/RequestMessageHandlerWrapperModel.cs:6]: This must have an interesting origin story :smile:

**Review by martinothamar — CHANGES_REQUESTED** [review]
Thanks for your patience, sorry about the delay here!

A couple of comments, otherwise this looks awesome :1st_place_medal: 
I must admit I don't understand all these Roslyn APIs related to incremental compilation, but you bring benchmarks (which I ran) and tests so LGTM :smile: 


``` ini

BenchmarkDotNet=v0.13.1, OS=pop 22.04
AMD Ryzen 5 5600X, 1 CPU, 12 logical and 6 physical cores
.NET SDK=6.0.417
  [Host] : .NET 6.0.25 (6.0.2523.51912), X64 RyuJIT

Job=InProcess  Toolchain=InProcessEmitToolchain  

```
|       Method |      Mean |    Error |   StdDev |    Gen 0 |    Gen 1 | Allocated |
|------------- |----------:|---------:|---------:|---------:|---------:|----------:|
|       Cached |  13.69 ms | 0.053 ms | 0.047 ms | 265.6250 |  78.1250 |     22 MB |
|      Compile |  14.89 ms | 0.056 ms | 0.049 ms | 265.6250 | 265.6250 |     23 MB |
|  LargeCached |  87.59 ms | 0.327 ms | 0.290 ms | 333.3333 | 166.6667 |     38 MB |
| LargeCompile | 154.20 ms | 0.807 ms | 0.755 ms | 750.0000 | 250.0000 |     77 MB |

**Comment by TimothyMakkison** [comment]
Thanks for the review, I'll resolve the issues when I have more time over the weekend 👍 

> I must admit I don't understand all these Roslyn APIs related to incremental compilation

TLDR, each step in a incremental generator (`Select`, `Where`, `SelectMany`), caches the input and output of the previous runs step. By using a custom comparer or object implementing `IEquality` we can use this behaviour to skip a section of the pipeline if the input object is the same as the last run.

In roslyn equivalent syntax (ie `SyntaxNode`, `InvocationExpressionSyntax`) is the same between runs whereas each compilation object will always be different between runs (`ITypeSymbol`, `ISymbol`, `Compilation`) and should not be cached for both performance and logic reasons. 

In our case I created a parse step that produces a cacheable object, this lets us skip the output stage.

**Comment by martinothamar** [comment]
Thanks! Makes sense

File: src/Mediator.SourceGenerator.Implementation/Models/RequestMessageHandlerWrapperModel.cs (hunk 3 above)
Comment (TimothyMakkison) [reply src/Mediator.SourceGenerator.Implementation/Models/RequestMessageHandlerWrapperModel.cs:6]: I'm not allowed to delete this file, I have no idea why not...

image

Comment (TimothyMakkison) [reply src/Mediator.SourceGenerator.Implementation/Models/RequestMessageHandlerWrapperModel.cs]: I was able to delete this but I'm not able to update`RequestMessageHandlerWrapper.cs` to this filename. I have no idea what voodoo caused this.

File: src/Mediator.SourceGenerator.Implementation/Analysis/RequestMessageHandlerWrapper.cs (hunk 2 above)
Comment (TimothyMakkison) [reply src/Mediator.SourceGenerator.Implementation/Analysis/RequestMessageHandlerWrapper.cs:13]: Not sure either. Rereading the code I'm wondering if using the immutable models in the mutable code is a good idea as it makes mistakes like this common.

File: src/Mediator.SourceGenerator.Implementation/Analysis/CompilationAnalyzer.cs (hunk 1 above)
Comment (TimothyMakkison) [reply src/Mediator.SourceGenerator.Implementation/Analysis/CompilationAnalyzer.cs:110]: Good catch, suprised a unit test didn't catch this.

**Comment by TimothyMakkison** [comment]
Thanks for the review, I think I've made all the relevant changes 👍 

- I opted to use your approach and moved `TryParseConfiguration` to before the array declaration
- `RequestMessageHandlerWrapper` is still a little janky. My ide doesn't show linting if the file is in the `Models` folder, but will lint and show references when moved out. Still no idea why this is happening. It occurs on different computers with fresh installs 😩
- I'm a little conversned that the tests didn't fail when I broke diagnostics or namespace configuration. I'm planning on adding some tests to `SnapshotTests`. Should I add them in this pr or a different one? This pr is getting a little bloated.

**Review by martinothamar — APPROVED** [review]
finally merging this, thanks for the great work! :smile:
